packages/neptune-extension/src/common/components/upload-modal/UploadForm.tsx

```diff
@@ -0,0 +1,185 @@
+import React from 'react';
+import { useDispatch, useSelector } from 'react-redux';
+
+import { NotebookDTO } from 'generated/leaderboard-client/src/models';
+
+import {
+  PlatformNotebook,
+  PlatformNotebookMetadata,
+} from 'types/platform';
+
+import {
+  uploadNotebook,
+  uploadCheckpoint,
+  getNotebook,
+} from 'common/utils/upload';
+
+import { PROJECT_LOCAL_STORAGE_KEY } from 'common/utils/localStorage';
+
+import Modal from 'common/components/modal/Modal';
+import ProjectInput from 'common/components/input/ProjectInput';
+
+import { getConfigurationState } from 'common/state/configuration/selectors';
+import { getNotebookState } from 'common/state/notebook/selectors'
+import { setNotebook } from 'common/state/notebook/actions'
+
+interface UploadFormProps {
+  platformNotebook: PlatformNotebook
+  onClose: () => void
+}
+
+type UploadMode = 'notebook' | 'checkpoint'
+
+const UploadForm: React.FC<UploadFormProps> = ({
+  platformNotebook,
+  onClose,
+}) => {
+  const [ metadata ] = React.useState(() => platformNotebook.getMetadata());
+
+  const dispatch = useDispatch()
+  const { notebook } = useSelector(getNotebookState)
+  const { inferredUsername } = useSelector(getConfigurationState)
+
+  const noAccess = !notebook && metadata.notebookId
+  const ownerChanged = notebook && notebook.owner !== inferredUsername
+  const pathChanged = notebook && notebook.path !== metadata.path
+
+  const canUploadCheckpoint = notebook && !ownerChanged
+
+  const [ loading, setLoading ] = React.useState(false);
+
+  const [ mode, setMode ] = React.useState<UploadMode>(canUploadCheckpoint ? 'checkpoint' : 'notebook');
+
+  const [ projectId, setProjectId ] = React.useState(() => {
+    return !!notebook
+      ? notebook.projectId
+      : window.localStorage.getItem(PROJECT_LOCAL_STORAGE_KEY) || ''
+  });
+  const [ name, setName ] = React.useState('');
+  const [ description, setDescription ] = React.useState('');
+
+  function changeMode(newMode: UploadMode) {
+    if (newMode === 'checkpoint') {
+      setProjectId((notebook as NotebookDTO).projectId);
+    }
+
+    setMode(newMode);
+  }
+
+  async function handleSubmit() {
+    setLoading(true);
+
+    const content = await platformNotebook.getContent();
+
+    const checkpointMeta = {
+      path: metadata.path,
+      name,
+      description,
+    };
+
+    if (mode === 'notebook') {
+      const notebookMeta = await uploadNotebook(projectId, checkpointMeta, content);
+
+      await platformNotebook.saveNotebookId(notebookMeta.id);
+
+      const notebook = await getNotebook(notebookMeta.id)
+      dispatch(setNotebook(notebook))
+    }
+    else {
+      await uploadCheckpoint(metadata.notebookId as string, checkpointMeta, content);
+
+      if (pathChanged) {
+        // Reload notebook
+        const notebook = await getNotebook(metadata.notebookId as string)
```

Comment:
I think the logic in here could be simplified.

You have:
- getNotebook (which fetches notebook) /Action
- setNotebook (which stores notebook in redux) /Action
- loadInitialNotebook (which fetches notebook - once) /Effect
- api calls in utils /Promise

This makes, so that a lot of logic is repeated and must be micro managed in this component (calling getNotebook on changes etc)

I think what you need are:
- fetchNotebook /Action
- createNotebook /Action
- createCheckpoint /Action
- fetchNotebook /Effect

fetchNotebook  has 3 actions:
- request(id) 
  If has no id, it dispatches fetch-failure
  If API fetch returns notebook dispatches fetch-success
  If API fetch returns ex dispatches fetch-error
- success - on this action redux stores the notebook
- failure - on this action redux clears stored notebook

createNotebook has 3 actions:
- request()
  calls API, dispatches success/failure. On this action it sets uploadStatus to pending
- success - on this action redux sets uploadStatus = success. In this action, we dispatch fetchNotebook.
- failure - on this action redux sets uploadStatus = failue.

createCheckpoint - analogically to createNotebook.

In UploadModal you listen to change of uploadStatus from redux.
If it changes from non-success to success - you stores id in platform and close modal
If uploadStatus = pending - you show spinner in modal
In App you useEffect fetchNotebook and it is depending on platform.notebook.id change. So its refetching Notebook when you set notebookId and also fetches initial notebook to the app.

IMHO it would largely simplify UploadModal

Reply:
If I understand you correctly, you propose to move the logic of reloading a notebook to a separate hook that listens to changes in redux ? Couple of points to make here:
1. For me explicit code is better, you have actions that happen one after another and it's easier to understand the flow.
2. Reload happens also when a path changes and this is stored only in platform at this moment. In that case notebook id does not change.
3. Here notebook reload is 2 lines of code, times two conditions it gives you 4 lines, both with different notebook id. I can check whether `mode === 'notebook'` or path changed and have only one reload, but then I have to store the current notebook is somewhere. Is it really worth it ?

Beside that writing an implicit logic that you seem to propose requires at least in my opinion an event system which redux is not. I know it is still possible with redux and I would also resort to such programming given a huge benefit, but in this case I just cannot see the simplification that you promise. Maybe a code snippet would help so I can spot it ?

Reply:
I suggest moving almost all of logic here to redux actions.

1. I disagree - it creates a lot of unnecessary repetition
``
await getNotebook(x) <-- you dont handle errors here
dispatch(setNotebook(x)

vs

dispatch(fetchNotebook()) 
``
2. I would reload each time after successful upload. I belive other notebook fields from API can be changed after upload so its safe to reaload it anyway.

3. If I understand correctly you'd prefer logic:

 1. Load initial notebook using effect
 2. At the end of UploadCheckpointSuccess action - fetchNotebook
 3. At the end of UploadNotebookSuccess action - fetchNotebok

as long as we remove getNotebook/setNotebook and introduce other actions. I think I'm ok with this.
__________________________________

I was thinking it would look like this
```
// inside render
React.useEffect((uploadStatus)) {
 if (uploadStatus !== Success) return;
 if (mode === notebook) {
   platform.setNotebookId(notebookId)
 } 
 closeModal()
 dispatch(resetUploadStatus) //to null
}

React.useEffect((metadataId) => {
      if (!metadata.notebokoId)  dispatch(fetchNotebookFailed(e))
      dispatch(fetchNotebook({id: metadataId}))
  }, [platformNotebook.getMetadata()])

<LoadingIcon visible={uploadStatus === 'pending'} />

handleSubmit() {
	  const content = await platformNotebook.getContent();
    const checkpointMeta = {
      path: metadata.path,
      name,
      description,
    };
 }

 if (mode === 'notebook') {
      dispatch(uploadNotebook({projectId, metadata: checkpointMeta, content}))
    } else {
      dispatch(uploadCheckpoint({notebookId, metadata: checkpointMeta, content});
    }
}

// Actions
uploadNotebook({projectId, metadata, content}) {
   try {
     await apiCall
     dispatch(uploadNotebookSuccess())
   } catch (e) {
     dispatch(uploadNotebookError(e))
   }
}

uploadCheckpoint({notebookId, metadata, content}) {
   try {
     await apiCall
     dispatch(uploadCheckpointSuccess())
     dispatch(fetchNotebook())
   } catch (e) {
     dispatch(uploadCheckpointError(e))
   }
}


```